File: src/main/core/modpack.js
'use strict';

/**
 * Сборка модов в файле — «поделиться сборкой», как экспорт профиля
 * в r2modman или .mrpack в Modrinth (1.10).
 *
 * Внутри не сами моды, а их номера в каталоге: файл весит пару килобайт,
 * и у друга ModHub скачает всё из тех же мест, откуда качали вы.
 * Моды, поставленные из файла, в каталоге не найти — они попадают в сборку
 * только для списка и при импорте честно показываются как «поставьте сами».
 */

const FORMAT = 'modhub-pack';
const VERSION = 1;
const MAX_MODS = 500;

function str(value, max = 200) {
  return typeof value === 'string' ? value.slice(0, max) : '';
}

/**
 * @param {object} game адаптер
 * @param {object[]} records записи реестра
 * @param {{name?: string, now?: number, app?: string}} [meta]
 */
function buildPack(game, records, meta = {}) {
  return {
    format: FORMAT,
    version: VERSION,
    app: meta.app ?? '',
    game: game.id,
    gameName: game.name,
    name: str(meta.name, 60) || game.name,
    createdAt: new Date(meta.now ?? Date.now()).toISOString(),
    mods: (records ?? [])
      .filter((r) => !r.missing)
      // Зависимости поставятся сами вслед за модом, который их попросил.
      .filter((r) => !r.requestedBy)
      .map((r) => ({
        id: r.id,
        name: r.name,
        version: r.version ?? '',
        source: r.source ?? 'file',
        url: r.url ?? null,
        enabled: r.enabled !== false,
      })),
  };
}

/**
 * Проверяет и нормализует чужой файл сборки. Ничему в нём не верим на слово:
 * файл мог прийти откуда угодно.
 * @returns {{game: string, name: string, mods: object[]}}
 */
function parsePack(text) {
  let data;
  try {
    data = JSON.parse(String(text ?? ''));
  } catch {
    throw Object.assign(new Error('not json'), { code: 'PACK_FORMAT' });
  }
  if (!data || data.format !== FORMAT || typeof data.game !== 'string' || !Array.isArray(data.mods)) {
    throw Object.assign(new Error('not a modhub pack'), { code: 'PACK_FORMAT' });
  }
  if (Number(data.version) > VERSION) {
    throw Object.assign(new Error('newer pack'), { code: 'PACK_NEWER' });
  }
  const mods = data.mods
    .slice(0, MAX_MODS)
    .filter((m) => m && typeof m.id === 'string' && m.id.length <= 200)
    .map((m) => ({
      id: m.id,
      name: str(m.name) || m.id,
      version: str(m.version, 40),
      source: str(m.source, 20) || 'file',
      url: /^https:\/\//i.test(m.url ?? '') ? str(m.url, 500) : null,
      enabled: m.enabled !== false,
    }));
  return { game: data.game, gameName: str(data.gameName, 80), name: str(data.name, 60), mods };
}

module.exports = { buildPack, parsePack, FORMAT };

File: desktop/ModLaunch/Features/ModPack.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using ModLaunch.Core;
using ModLaunch.Games;
using ModLaunch.Mods;

namespace ModLaunch.Features;

public sealed record PackMod(string Id, string Name, string Version, string Source, string? Url, bool Enabled);
public sealed record Pack(string Game, string GameName, string Name, List<PackMod> Mods);

/// <summary>Сборка в файле (.modhub.json) — тот же формат, что у 3.x, файлы совместимы.</summary>
public static class ModPack
{
    const string Format = "modhub-pack";
    const int Version = 1;

    static string Str(JsonNode? n, string key, int max = 200)
    {
        var s = n.Str(key) ?? "";
        return s.Length > max ? s[..max] : s;
    }

    public static string Build(GameDef game, ModRegistry registry, string name)
    {
        var mods = registry.List().Where(r => !r.Bool("missing") && r.Str("requestedBy") is null)
            .Select(r => (JsonNode)new JsonObject
            {
                ["id"] = r.Str("id"),
                ["name"] = r.Str("name"),
                ["version"] = r.Str("version") ?? "",
                ["source"] = r.Str("source") ?? "file",
                ["url"] = r.Str("url"),
                ["enabled"] = r.Bool("enabled", true),
            }).ToArray();
        var pack = new JsonObject
        {
            ["format"] = Format,
            ["version"] = Version,
            ["app"] = Http.Version,
            ["game"] = game.Id,
            ["gameName"] = game.Name,
            ["name"] = string.IsNullOrWhiteSpace(name) ? game.Name : name.Trim(),
            ["createdAt"] = DateTime.UtcNow.ToString("o"),
            ["mods"] = new JsonArray(mods),
        };
        return pack.ToJsonString(new JsonSerializerOptions { WriteIndented = true });
    }

    public static int Count(string json) => JsonNode.Parse(json).Arr("mods").Count;

    public static Pack Parse(string text)
    {
        JsonNode? data;
        try { data = JsonNode.Parse(text); }
        catch { throw new InvalidDataException(I18n.T("err.PACK_FORMAT")); }
        if (data.Str("format") != Format || data.Str("game") is null || data?["mods"] is not JsonArray list)
            throw new InvalidDataException(I18n.T("err.PACK_FORMAT"));
        if (data.Long("version") > Version) throw new InvalidDataException(I18n.T("err.PACK_NEWER"));
        var mods = list.Take(500)
            .Where(m => m.Str("id") is { Length: > 0 and <= 200 })
            .Select(m => new PackMod(
                m.Str("id")!,
                Str(m, "name") is { Length: > 0 } n ? n : m.Str("id")!,
                Str(m, "version", 40),
                Str(m, "source", 20) is { Length: > 0 } s ? s : "file",
                (m.Str("url") ?? "").StartsWith("https://", StringComparison.OrdinalIgnoreCase) ? Str(m, "url", 500) : null,
                m.Bool("enabled", true)))
            .ToList();
        return new Pack(data.Str("game")!, Str(data, "gameName", 80), Str(data, "name", 60), mods);
    }
}

/// <summary>Сравнение версий модов: 1.10.0 новее 1.9.2, «v» в начале не мешает.</summary>
public static class Versions
{
    static double[]? Parts(string? version)
    {
        var clean = (version ?? "").Trim();
        if (clean.Length > 1 && (clean[0] is 'v' or 'V') && char.IsDigit(clean[1])) clean = clean[1..];
        var main = clean.Split('-', '+', ' ')[0];
        var parts = main.Split('.');
        var result = new double[parts.Length];
        for (var i = 0; i < parts.Length; i++)
            if (!double.TryParse(parts[i], System.Globalization.NumberStyles.None, null, out result[i])) return null;
        return result;
    }

    public static int Compare(string? a, string? b)
    {
        var x = Parts(a);
        var y = Parts(b);
        if (x is null || y is null) return (a ?? "").Trim() == (b ?? "").Trim() ? 0 : 1;
        for (var i = 0; i < Math.Max(x.Length, y.Length); i++)
        {
            var d = (i < x.Length ? x[i] : 0) - (i < y.Length ? y[i] : 0);
            if (d != 0) return d > 0 ? 1 : -1;
        }
        return 0;
    }

    public static bool IsNewer(string? latest, string? installed) =>
        !string.IsNullOrEmpty(latest) && !string.IsNullOrEmpty(installed) && Compare(latest, installed) > 0;
}
